### ISSUE TITLE
modules within modules

### ISSUE THREAD

**Issue by mighty-pirate** [revel/revel#1310]
I have a revel application and 2 modules. I could add the 2 modules to the application but I was wondering if it was possible to make 1 module the dependency of another so when I come to reuse this module, it can just be 'dropped in'. Am I correct in thinking that config files are not used for modules?

I can build a workaround that reads a module config file that lists required dependencies, and loads them on boot, but didn't want to to do this if something was already possible within revel itself.

**Comment by notzippy** [comment]
**tl;dr** - no there is not something currently in revel that does this.

That is something I have wanted to do have a `module.conf`, so a module could append its own dependent modules (similar to what the routes files does). 
This becomes a rather intriguing project involving both the revel/cmd package and the revel/revel, rules about what can a module.conf set (ie can it specify a database connection, and how should the module configuration properties be mapped onto revel.Config object?). The other thing I was thinking about if the `module.conf` could define a prefix for the template path - or maybe the template paths prefix could be defined by the name of the module that was imported... All interesting topics :-P If you want to take a crack at implementing this I think it would be a great addition to Revel. If you want to bounce some ideas off of me I am on https://gitter.im/revel/community

**Comment by mighty-pirate** [comment]
I'd definitely like to try building something into revel but it might not be for a few weeks yet.

It could be worth namespacing the config per module, but that would mean that a module cannot override the base config at all. Though if overriding is possible, it could get messy with different modules conflicting with and overriding one another, so it would have to be used with caution. 

It would probably make sense to have a base config file within the module, for the defaults called 'module.conf'. Then within the project using the module, use a config file for any changes to the defaults. This would be named using the identifer that was specified when importing the module so for `modules.pages = github.com/username/module-name` the config would be called 'pages.conf'.

The same would then apply to modules using modules. So in the parent module.conf you would have `modules.child = github.com/username/child-module-name` which would read the default module.conf from the child module and then the child.conf from the parent module.

The example I have given above probably wouldn't work very well unless config was namespaced and only accessible at the level the module was imported, and the module itself.  

Another issue which could arise is an infinite loop of importing modules. It could be solved with some logic that stops a module from importing a parent/grandparent/etc. This doesn't have to be included though, as a warning may suffice.
